experimental/codepointtrie/src/lib.rs

```diff
@@ -0,0 +1,338 @@
+
+const CODEPOINTTRIE_SMALL_LIMIT: u32 = 0x10000;
+
+const CODEPOINTTRIE_SMALL_INDEX_LENGTH: u32 =
+    CODEPOINTTRIE_SMALL_LIMIT >> CODEPOINTTRIE_FAST_TYPE_SHIFT;
+
+/// Shift size for getting the index-3 table offset.
+const CODEPOINTTRIE_SHIFT_3: u32 = 4;
+
+/// Shift size for getting the index-2 table offset.
+const CODEPOINTTRIE_SHIFT_2: u32 = 5 + CODEPOINTTRIE_SHIFT_3;
+
+/// Shift size for getting the index-1 table offset.
+const CODEPOINTTRIE_SHIFT_1: u32 = 5 + CODEPOINTTRIE_SHIFT_2;
+
+/// Difference between two shift sizes,
+///  for getting an index-2 offset from an index-3 offset. 5=9-4
+const CODEPOINTTRIE_SHIFT_2_3: u32 = CODEPOINTTRIE_SHIFT_2 - CODEPOINTTRIE_SHIFT_3;
+
+/// Difference between two shift sizes,
+/// for getting an index-1 offset from an index-2 offset. 5=14-9
+const CODEPOINTTRIE_SHIFT_1_2: u32 = CODEPOINTTRIE_SHIFT_1 - CODEPOINTTRIE_SHIFT_2;
+
+/// Number of index-1 entries for the BMP. (4)
+/// This part of the index-1 table is omitted from the serialized form.
+const CODEPOINTTRIE_OMITTED_BMP_INDEX_1_LENGTH: u32 = 0x10000 >> CODEPOINTTRIE_SHIFT_1;
+
+/// Number of entries in an index-2 block. 32=0x20
+const CODEPOINTTRIE_INDEX_2_BLOCK_LENGTH: u32 = 1 << CODEPOINTTRIE_SHIFT_2;
+
+/// Mask for getting the lower bits for the in-index-2-block offset.
+const CODEPOINTTRIE_INDEX_2_MASK: u32 = CODEPOINTTRIE_INDEX_2_BLOCK_LENGTH - 1;
+
+/// Number of code points per index-2 table entry. 512=0x200
+const CODEPOINTTRIE_CP_PER_INDEX_2_ENTRY: u32 = 1 << CODEPOINTTRIE_SHIFT_2;
+
+/// Number of entries in an index-3 block. 32=0x20
+const CODEPOINTTRIE_INDEX_3_BLOCK_LENGTH: u32 = 1 << CODEPOINTTRIE_SHIFT_2_3;
+
+/// Mask for getting the lower bits for the in-index-3-block offset.
+const CODEPOINTTRIE_INDEX_3_MASK: u32 = CODEPOINTTRIE_INDEX_3_BLOCK_LENGTH - 1;
+
+/// Number of entries in a small data block. 16=0x10
+const CODEPOINTTRIE_SMALL_DATA_BLOCK_LENGTH: u32 = 1 << CODEPOINTTRIE_SHIFT_3;
+
+/// Mask for getting the lower bits for the in-small-data-block offset.
+const CODEPOINTTRIE_SMALL_DATA_MASK: u32 = CODEPOINTTRIE_SMALL_DATA_BLOCK_LENGTH - 1;
+
+/// The width of the elements in the data array of a CodePointTrie.
+/// See UCPTrieValueWidth in ICU4C.
+enum CodePointTrieValueWidth {
+    BitsAny = -1,
+    Bits16 = 0,
+    Bits32 = 1,
+    Bits8 = 2,
+}
+
+/// The type of trie represents whether the trie has an optimization that
+/// would make it small or fast.
+/// See UCPTrieType in ICU4C.
+#[derive(PartialEq)]
+enum CodePointTrieType {
+    Any = -1,
+    Fast = 0,
+    Small = 1,
+}
+
+struct CodePointTrieData<'trie> {
+    // void: bool,  // Do we need an equivalent to the `void *ptr0` field in
+    // UCPTrieData? Is it derivative of a None value for the other fields? Is
+    // there a performance / convenience to having it?
+    data_8_bit: Option<&'trie [u8]>,
+    data_16_bit: Option<&'trie [u16]>,
+    data_32_bit: Option<&'trie [u32]>,
+}
+
+struct CodePointTrie<'trie> {
```

Comment:
templatize/parameterize struct by trie type and value width

Reply:
CodePointTrie has variations along two axes: type (fast vs. small) which affects the index computation, and value width (currently 8/16/32 bits) which affects the storage of and access to the data array. This makes for 6 combinations where some performance-critical code paths differ. In the future, there could be more, possibly for an additional value width.

In Java, this is handled via polymorphism. Class CodePointTrie has some intermediate and 6 final subclasses. The base class uses some generic types and implementations, and the subclasses are fixed and specialized. Simple and efficient.

In C, I use a union for the data pointers. C functions check the type and width each time using `switch` statements. Higher-performance code uses macros for type=fast access, combined with orthogonal macros for the value width. This yields maximum performance with statically chosen inline code, but no type safety.

What's the appropriate approach for Rust? It would be easy to implement it like in the C functions, with `switch` dispatch each time. It would be nice to make it more efficient.

Rust "traits" are roughly like Java interfaces, right? We could model the Rust API after the Java CodePointMap+CodePointTrie base classes, with 6 specific implementation classes that can share most of their code via calling internal functions. Instantiating a trie for given data would return the specific implementation type, like in Java. (See [CodePointTrie](https://github.com/unicode-org/icu/blob/main/icu4j/main/classes/core/src/com/ibm/icu/util/CodePointTrie.java).fromBinary(type, valueWidth, bytes), especially lines 242..260.)

Reply:
The Rusty style would be to make CodePointTrie be a single class (struct) with two type parameters, one for fast/small and the other for 8/16/32, and the methods can be specialized based on the type parameters.